proof/refine/RISCV64/IpcCancel_R.thy

```diff
@@ -78,261 +121,305 @@ locale delete_one_conc_pre =
+    "\<lbrace>valid_inQ_queues and valid_objs'\<rbrace> cteDeleteOne sl \<lbrace>\<lambda>rv. valid_inQ_queues\<rbrace>"
   assumes delete_one_ksCurDomain:
     "\<And>P. \<lbrace>\<lambda>s. P (ksCurDomain s)\<rbrace> cteDeleteOne sl \<lbrace>\<lambda>_ s. P (ksCurDomain s)\<rbrace>"
   assumes delete_one_tcbDomain_obj_at':
     "\<And>P. \<lbrace>obj_at' (\<lambda>tcb. P (tcbDomain tcb)) t'\<rbrace> cteDeleteOne slot \<lbrace>\<lambda>_. obj_at' (\<lambda>tcb. P (tcbDomain tcb)) t'\<rbrace>"
 
-lemma (in delete_one_conc_pre) cancelIPC_simple[wp]:
-  "\<lbrace>\<top>\<rbrace> cancelIPC t \<lbrace>\<lambda>rv. st_tcb_at' simple' t\<rbrace>"
-  apply (simp add: cancelIPC_def Let_def getThreadReplySlot_def
-             cong: Structures_H.thread_state.case_cong list.case_cong)
-  apply (rule hoare_seq_ext [OF _ gts_sp'])
-  apply (rule hoare_pre)
-   apply (wpc
-           | wp sts_st_tcb_at'_cases hoare_vcg_conj_lift
-                hoare_vcg_const_imp_lift delete_one_st_tcb_at
-                threadSet_pred_tcb_no_state
-                hoare_strengthen_post [OF cancelSignal_simple]
-           | simp add: o_def if_fun_split
-           | rule hoare_drop_imps
-           | clarsimp elim!: pred_tcb'_weakenE)+
-  apply (auto simp: pred_tcb_at'
-             elim!: pred_tcb'_weakenE)
-  done
-
-lemma (in delete_one_conc_pre) cancelIPC_st_tcb_at':
-  "\<lbrace>st_tcb_at' P t' and K (t \<noteq> t')\<rbrace>
-     cancelIPC t
-   \<lbrace>\<lambda>rv. st_tcb_at' P t'\<rbrace>"
-  apply (simp add: cancelIPC_def Let_def getThreadReplySlot_def locateSlot_conv
-                   capHasProperty_def isCap_simps)
-  apply (wp sts_pred_tcb_neq' hoare_drop_imps delete_one_reply_st_tcb_at
-       | wpc | clarsimp)+
-          apply (wp getCTE_wp | clarsimp)+
-         apply (wp hoare_vcg_ex_lift threadSet_cte_wp_at' hoare_vcg_imp_lift
-                   cancelSignal_pred_tcb_at' sts_pred_tcb_neq' getEndpoint_wp gts_wp'
-                   threadSet_pred_tcb_no_state
-              | wpc | clarsimp)+
-  apply (auto simp: cte_wp_at_ctes_of isCap_simps)
+lemma cancelSignal_st_tcb_at':
+  "\<lbrace>K (P Inactive)\<rbrace>
+   cancelSignal t ntfn
+   \<lbrace>\<lambda>_. st_tcb_at' P t\<rbrace>"
+  unfolding cancelSignal_def Let_def
+  apply (rule hoare_gen_asm_single)
+  supply getNotification_inv[wp del] set_ntfn'.get_inv[wp del]
+  apply (wpsimp wp: setThreadState_st_tcb_at'_cases
+                    hoare_drop_imp[where R="\<lambda>rv _. delete t (f rv) = []" for f]
+                    hoare_drop_imp[where R="\<lambda>rv _. \<exists>a b. delete t (f rv) = a # b" for f])
   done
 
 context begin interpretation Arch .
 crunch typ_at'[wp]: emptySlot "\<lambda>s. P (typ_at' T p s)"
 end
 
-crunch tcb_at'[wp]: cancelSignal "tcb_at' t"
-  (wp: crunch_wps simp: crunch_simps)
+sublocale delete_one_conc_pre < delete_one: typ_at_all_props' "cteDeleteOne slot"
+  by typ_at_props'
 
-context delete_one_conc_pre
-begin
+declare if_weak_cong [cong]
```

Comment:
This is one of these problematic ones, probably auto-inserted from the merge. `if_weak_cong` is the default, so in a sense this should be harmless, but it means that the context expects `if_cong` to be set as the cong rule for if at this point. If this line is producing a warning that it's duplicate, then it should be removed. If the line does not produce a warning, we should eventually eliminate the global declare further up in the tree somewhere that makes this one necessary. But not in this PR. For now, if we can move it up to the top of this theory and add a FIXME, that would be good already. If proofs break because of the move, you might have to add a local declaration for `if_cong[cong]`

Reply:
Moved to the top of the file, with a `FIXME RT: remove`